src/clients/rust/src/time_based_id.rs

```diff
@@ -0,0 +1,314 @@
+            .map(|d| d.as_millis())
+            .unwrap_or(0);
+
+        match next_ms_since_epoch.cmp(&previous_ms_since_epoch) {
+            Ordering::Greater => {
+                // The new time is greater than previous.
+                // Use it and choose a new random number.
+                TbidGenerator {
+                    ms_since_epoch: next_ms_since_epoch,
+                    random: random_u80(),
+                }
+            }
+            Ordering::Equal | Ordering::Less => {
+                // The new time is equal or less than previous.
+                // Use the old time and increment the random number.
+                match self.next_random() {
+                    Some(next_random) => TbidGenerator {
+                        ms_since_epoch: previous_ms_since_epoch,
+                        random: next_random,
+                    },
+                    None => {
+                        // Difficult case.
+                        //
+                        // `self.random` would overflow a `u80`.
+                        //
+                        // It _seems_ to be extremely unlikely: it requires many
+                        // ids to be generated in a single millisecond, and also
+                        // for the initial value of `self.random` to be large;
+                        // though consider that the clock may be broken and a
+                        // "millisecond" could last forever.
+                        //
+                        // The ULID spec says if the randomness overflows 80
+                        // bits then it is an error, but we instead manually
+                        // carry the bit from the randomness to the timestamp,
+                        // moving time forward by 1ms, and generate a new random
+                        // number. Generated ids will be ahead of the wall
+                        // clock by 1 ms until time catches up.
+                        TbidGenerator {
+                            // There is too much resolution in a u128 for this to ever overflow.
+                            ms_since_epoch: previous_ms_since_epoch
+                                .checked_add(1)
+                                .expect("impossible overflow"),
+                            random: random_u80(),
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn next_random(&self) -> Option<u128> {
+        assert!(is_u80(self.random));
+        if self.random == U80_MASK {
+            None
+        } else {
+            Some(self.random + 1)
+        }
+    }
+}
+
+const U80_MASK: u128 = 0x_FFFF_FFFF_FFFF_FFFF_FFFF;
+
+fn is_u80(val: u128) -> bool {
+    val <= U80_MASK
+}
+
+fn random_u80() -> u128 {
+    random_u128() & U80_MASK
+}
+
+fn random_u128() -> u128 {
+    let a = random_u64();
+    let b = random_u64();
+    let a = a as u128;
+    let b = b as u128;
+    a | b << 64
```

Comment:
```suggestion
    a | (b << 64)
```

so that the reader doesn't have to wander about the order of operators in Rust.